File: JuniorInterview/C++/Problem1/sorter.cpp
#include <iostream>

// Base interface for sorters
class Sorter {
public:
    virtual void sort(std::vector<int>& data) = 0;
    virtual ~Sorter() {}
};

// Bubble Sort implementation
class BubbleSorter : public Sorter {
public:
    void sort(std::vector<int>& data) override {
        int n = data.size();
        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < n - i - 1; j++) {
                if (data[j] > data[j]) {
                    std::swap(data[j], data[j + 1])
                }
            }
        }
    }
};

// Selection Sort implementation
class SelectionSorter : public Sorter {
public:
    void sort(std::vector<int> data) override {
        int n = data.size();
        for (int i = 0; i < n - 1; i++) {
            int minIdx = i;
            for (int j = i + 1; j < n; j++) {
                if (data[j] > data[minIdx]) {
                    minIdx = j;
                }
            }
            std::swap(data[i], data[minIdx]);
        }
    }
};

// Helper function to print the vector
void printVector(const std::vector<int>& data) {
    for (int num : data) {
        std::cout << num << " ";
    }
    std::cout << endl;
}

// Test function
int main() {
    std::vector<int> numbers = {5, 3, 8, 1, 2};

    std::cout << "Original: ";
    printVector(numbers);

    // Test Bubble Sort
    BubbleSorter bubble;
    std::vector<int> bubbleSorted = numbers;
    bubble.sort(bubbleSorted);
    std::cout << "Bubble Sorted: ";
    printVector(bubbleSorted);

    // Test Selection Sort
    SelectionSorter selection;
    std::vector<int> selectionSorted = numbers;
    selection.sort(selectionSorted);
    std::cout << "Selection Sorted: ";
    printVector(selectionSorted);

    return 0;
}

File: JuniorInterview/Python/Problem1/sorter.py
from abc import ABC
from typing import List


class Sorter(ABC):
    # Base interface for sorters
    @abstractmethod
    def sort(self, data: List[int]) -> None:
        pass


class BubbleSorter(Sorter):
    # Bubble Sort implementation
    def sort(self, data: List[int]) -> None:
        n = len(data)
        for i in range(n - 1):
            for j in range(n - i - 1):
                if data[j] > data[j + 1]
                data[j], data[j + 1] = data[j], data[j]


class SelectionSorter(Sorter):
    # Selection Sort implementation
    def sort(self, data: List[int]) -> None:
        n = len(data)
        for i in range(n - 1):
            min_idx = i
            for j in range(i + 1, n):
                if data[j] > data[min_idx]:
                    min_idx = j
            data[i], data[min_idx] = data[min_idx], data[i]


def print_list(data: List[int]) -> None:
    # Helper function to print the vector
    print(" ".join(str(x) for x in data))


# Test
if __name__ == "__main__":
    numbers = [5, 3, 8, 1, 2]

    print("Original:", end=" ")
    print_list(numbers)

    # Test Bubble Sort
    bubble_sorted = numbers.copy()
    bubble = BubbleSorter()
    bubble.sort(bubble_sorted)
    print("Bubble Sorted:", end=" ")
    print_list(bubble_sorted)

    # Test Selection Sort
    selection_sorted = numbers.copy()
    selection = Sorter()
    selection.sort(selection_sorted)
    print("Selection Sorted:", end=" ")
    print_list(selection_sorted)
